### PR TITLE
PoS Mempool Transaction Validation

### PR DESCRIPTION
**Pull request by AeonSw4n**

### CODE REVIEW

**Comment by AeonSw4n** [comment]
Current dependencies on/for this PR:
* feature/proof-of-stake
  * **PR #578** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/578" target="_blank">Graphite</a> 
    * **PR #563** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/563" target="_blank">Graphite</a> 
      * **PR #599** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/599" target="_blank">Graphite</a> 
        * **PR #600** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/600" target="_blank">Graphite</a>  👈
          * **PR #601** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/601" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/600?utm_source=stack-comment).

File: lib/validation_msg_deso_txn.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+		// Check that each output's amount is not bigger than the max as a
+		// sanity check.
+		if txout.AmountNanos > MaxNanos {
+			return errors.Wrapf(RuleErrorOutputExceedsMax, "ValidateTransactionSanityBalanceModel: Output amount %d "+
+				"exceeds max %d", txout.AmountNanos, MaxNanos)
+		}
+		// Check that this output doesn't overflow the total as a sanity
+		// check. This is frankly impossible since our maximum limit is
+		// not close to the max size of a uint64 but check it nevertheless.
+		if totalOutNanos >= math.MaxUint64-txout.AmountNanos {
```
Comment (lazynina) [inline lib/validation_msg_deso_txn.go:148]: This function never increases totalOutNanos. You can use the `SafeUint64` functions to do a safe add that throws an error if it overflows.

File: lib/validation_msg_deso_txn_test.go [hunk 2]
```diff
@@ -0,0 +1 @@
+package lib
```
Comment (lazynina) [inline lib/validation_msg_deso_txn_test.go:1]: What's going on in this file? If you're setting up for some future use, please leave a comment in here.

**Review by lazynina — CHANGES_REQUESTED** [review]

File: lib/validation_msg_deso_txn_test.go (hunk 2 above)
Comment (AeonSw4n) [reply lib/validation_msg_deso_txn_test.go]: Oh, this PR was a draft. We've got tests here now.

File: lib/validation_msg_deso_txn.go (hunk 1 above)
Comment (AeonSw4n) [reply lib/validation_msg_deso_txn.go]: Nice! Forgot `SafeUint64` existed. I copied this code from the BlockView connect logic. One difference is that the BlockView will error if `totalOutNanos + txout.AmountNanos == math.MaxUint64` while `SafeUint64` won't. Don't think it matters tho.

File: lib/network.go [hunk 3]
```diff
@@ -3779,6 +3779,24 @@ func (msg *MsgDeSoTxn) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+func (txn *MsgDeSoTxn) ComputeFeePerKB() (uint64, error) {
```
[the remark this answers is not in the document]
Comment (AeonSw4n) [reply lib/network.go]: Also added an explicit check that throws an error if Signature is nil.

File: lib/validation_deso_txn.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+		// Check that each output's amount is not bigger than the max as a
+		// sanity check.
+		if txout.AmountNanos > MaxNanos {
+			return errors.Wrapf(RuleErrorOutputExceedsMax, "ValidateDeSoTxnFormatBalanceModel: Output amount %d "+
+				"exceeds max %d", txout.AmountNanos, MaxNanos)
+		}
+		// Check that this output doesn't overflow the total as a sanity
+		// check. This is frankly impossible since our maximum limit is
+		// not close to the max size of a uint64 but check it nevertheless.
+		if _, err := SafeUint64().Add(totalOutNanos, txout.AmountNanos); err != nil {
```
Comment (lazynina) [inline lib/validation_deso_txn.go:139]: I think we want totalOutputNanos on the left side here. Otherwise, we never increase this.

**Review by lazynina — CHANGES_REQUESTED** [review]

**Review by lazynina — APPROVED** [review]

**Pull request merged**